pkg/lib/cockpit/dialog.tsx

```diff
@@ -0,0 +1,1543 @@
+        this.#update();
+        if (!await this.validate()) {
+            this.#action_running = false;
+            this.#update();
+            return;
+        }
+
+        try {
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
+        }
+
+        this.#action_running = false;
+        this.#update();
+    }
+
+    top_value(update_func?: ((val: V) => void) | undefined): DialogValue<V> {
+        return new DialogValue<V>(
+            this as DialogState<unknown>,
+            () => this.values,
+            (val) => {
+                debug("set", val);
+                if (this.#action_running) {
+                    debug("set denied");
+                    return;
+                }
+                this.values = val;
+                this.#update();
+                if (this.#online_validation)
+                    this.#trigger_validation();
+                if (update_func)
+                    update_func(val);
+            },
+            "");
+    }
+
+    value<K extends keyof V>(tag: K, update_func?: ((val: V[K]) => void) | undefined): DialogValue<V[K]> {
+        return this.top_value().sub(tag, update_func);
+    }
+}
+
+export class DialogError {
+    title: string;
+    details: React.ReactNode;
+
+    constructor(title: string, details: React.ReactNode) {
+        this.title = title;
+        this.details = details;
+    }
+
+    toString() {
+        return this.title + ": " + String(this.details);
+    }
+
+    static fromError(title: string, err: unknown) {
+        if (err && typeof err == "object" && "message" in err && typeof err.message == "string")
+            return new DialogError(title, err.message);
+        else
+            return new DialogError(title, String(err));
+    }
+}
+
+export function useDialogState<V extends object>(
+    init: () => V,
+    validate: (dlg: DialogState<V>) => void
+) : DialogState<V> {
+    const dlg = useObject(() => new DialogState(init(), validate), null, []);
+    useOn(dlg, "changed");
+    return dlg;
+}
+
+export function useDialogState_async<V extends object>(
+    init: () => Promise<V>,
+    validate: (dlg: DialogState<V>) => void
+) : null | DialogError | DialogState<V> {
+    const [dlg, setDlg] = useState<null | DialogError | DialogState<V>>(null);
+    useOn((dlg instanceof DialogError ? null : dlg), "changed");
+    useInit(async () => {
+        try {
+            setDlg(new DialogState<V>(await init(), validate));
+        } catch (ex) {
+            if (ex instanceof DialogError)
+                setDlg(ex);
+            else
+                setDlg(DialogError.fromError(_("Error during initialization"), ex));
+        }
+    });
+    return dlg;
+}
+
+// Common elements
+
+export function DialogErrorMessage<V>({
+    dialog,
+} : {
+    dialog: DialogState<V> | DialogError | null,
+}) {
+    const err = (!dialog || dialog instanceof DialogError) ? dialog : dialog.error;
```

Comment:
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22700-2dd5f8c0-20251219-103022-fedora-43-devel-other/Coverage/lcov/github-pr.diff.gcov.html)